server/src/main/java/io/crate/expression/symbol/Symbol.java

```diff
@@ -197,14 +197,19 @@ default Symbol cast(DataType<?> targetType, CastMode... modes) {
             return this;
         } else if (targetType.equals(valueType())) {
             return this;
-        } else if (ArrayType.unnest(targetType).equals(DataTypes.UNTYPED_OBJECT)
-                   && valueType().id() == targetType.id()) {
-            return this;
-        } else if (ArrayType.unnest(targetType).equals(DataTypes.NUMERIC)
-                   && valueType().id() == DataTypes.NUMERIC.id()) {
-            // Do not cast numerics to unscaled numerics because we do not want to loose precision + scale
-            return this;
+        } else {
+            DataType<?> innerTargetType = ArrayType.unnest(targetType);
+            if (innerTargetType.equals(DataTypes.UNTYPED_OBJECT) &&
+                innerTargetType.id() == ArrayType.unnest(valueType()).id() &&
```

Comment:
This ensures the inner type is checked, and if it does not match, we try to cast it.